File: BlackjackApp/Domain/Models/Player.swift
import Foundation

enum PlayerType: Sendable {
    case human
    case bot
}

// XP Rank thresholds (cumulative XP)
enum PlayerRank: String, CaseIterable, Sendable {
    case beginner  = "Débutant"
    case initiate  = "Initié"
    case pro       = "Pro"
    case legend    = "Légende"

    static func rank(for xp: Int) -> PlayerRank {
        switch xp {
        case 0..<500:    return .beginner
        case 500..<2000: return .initiate
        case 2000..<5000: return .pro
        default:          return .legend
        }
    }
}

@Observable
final class Player: Identifiable {
    let id: UUID        = UUID()
    let type: PlayerType
    var name: String
    var bankroll: Double

    // Up to 4 hands after splits (European BJ max split depth)
    var hands: [Hand]   = [Hand()]
    var activeHandIndex: Int = 0

    // Side bet placed this round
    var sideBetWager: SideBetWager?

    // XP / pedagogy (human player only)
    var xp: Int         = 0
    var playerRank: PlayerRank { PlayerRank.rank(for: xp) }

    // Seat position at the table (0-4)
    let seatIndex: Int

    var currentHand: Hand {
        get { hands[activeHandIndex] }
        set { hands[activeHandIndex] = newValue }
    }

    var bet: Double     = 0

    init(type: PlayerType, name: String, bankroll: Double, seatIndex: Int) {
        self.type      = type
        self.name      = name
        self.bankroll  = bankroll
        self.seatIndex = seatIndex
    }

    func resetForNewRound() {
        hands           = [Hand()]
        activeHandIndex = 0
        bet             = 0
        sideBetWager    = nil
    }
}

File: src/domain/models/player.ts
import { v4 as uuidv4 } from 'uuid'
import { Hand } from './hand'
import { SideBetWager } from './sideBet'

export enum PlayerType { HUMAN = 'human', BOT = 'bot' }

export enum PlayerRank {
  BEGINNER = 'Débutant',
  INITIATE = 'Initié',
  PRO      = 'Pro',
  LEGEND   = 'Légende',
}

export function xpToPlayerRank(xp: number): PlayerRank {
  if (xp < 500)  return PlayerRank.BEGINNER
  if (xp < 2000) return PlayerRank.INITIATE
  if (xp < 5000) return PlayerRank.PRO
  return PlayerRank.LEGEND
}

export interface Player {
  readonly id: string
  readonly type: PlayerType
  readonly seatIndex: number  // 0-4 for seats; -1 for dealer
  name: string
  bankroll: number
  hands: Hand[]
  activeHandIndex: number
  sideBetWager: SideBetWager | null
  xp: number
  bet: number
}

export function createPlayer(
  type: PlayerType, name: string, bankroll: number, seatIndex: number
): Player {
  return {
    id: uuidv4(),
    type,
    name,
    bankroll,
    seatIndex,
    hands: [new Hand()],
    activeHandIndex: 0,
    sideBetWager: null,
    xp: 0,
    bet: 0,
  }
}

export function currentHand(player: Player): Hand {
  const hand = player.hands[player.activeHandIndex]
  if (!hand) throw new Error(`activeHandIndex ${player.activeHandIndex} out of range`)
  return hand
}

export function resetPlayerForNewRound(player: Player): void {
  player.hands = [new Hand()]
  player.activeHandIndex = 0
  player.bet = 0
  player.sideBetWager = null
}

export function createDealer(): Player {
  return createPlayer(PlayerType.BOT, 'Dealer', Infinity, -1)
}
